internal/controlplane/handlers_repositories.go

```diff
@@ -308,29 +309,61 @@ func (s *Server) ListRemoteRepositoriesFromProvider(
 	entityCtx := engine.EntityFromContext(ctx)
 	projectID := entityCtx.Project.ID
 
-	provider, err := getProviderFromRequestOrDefault(ctx, s.store, in, projectID)
+	provs, err := listProvidersOrInferDefault(ctx, s.store, in, projectID)
 	if err != nil {
 		return nil, providerError(err)
 	}
 
-	zerolog.Ctx(ctx).Debug().
-		Str("provider", provider.Name).
-		Str("projectID", projectID.String()).
-		Msg("listing repositories")
-
-	pbOpts := []providers.ProviderBuilderOption{
-		providers.WithProviderMetrics(s.provMt),
-		providers.WithRestClientCache(s.restClientCache),
-	}
-	p, err := providers.GetProviderBuilder(ctx, provider, s.store, s.cryptoEngine, &s.cfg.Provider, pbOpts...)
-	if err != nil {
-		return nil, status.Errorf(codes.Internal, "cannot get provider builder: %v", err)
+	out := &pb.ListRemoteRepositoriesFromProviderResponse{
+		Results: []*pb.UpstreamRepositoryRef{},
 	}
 
-	if !p.Implements(db.ProviderTypeRepoLister) {
-		return nil, util.UserVisibleError(codes.Unimplemented, "provider does not implement repository listing")
+	failIfListerNotImplemented := len(provs) == 1
+
+	for _, provider := range provs {
+		zerolog.Ctx(ctx).Debug().
+			Str("provider", provider.Name).
+			Str("projectID", projectID.String()).
+			Msg("listing repositories")
+
+		pbOpts := []providers.ProviderBuilderOption{
+			providers.WithProviderMetrics(s.provMt),
+			providers.WithRestClientCache(s.restClientCache),
+		}
+		p, err := providers.GetProviderBuilder(ctx, provider, s.store, s.cryptoEngine, &s.cfg.Provider, pbOpts...)
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "cannot get provider builder: %v", err)
+		}
+
+		// check if the provider implements the repo lister interface
+		if failIfListerNotImplemented && !p.Implements(db.ProviderTypeRepoLister) {
+			return nil, util.UserVisibleError(codes.Unimplemented, "provider does not implement repository listing")
+		} else if !p.Implements(db.ProviderTypeRepoLister) {
+			// skip the provider if it does not implement the repo lister interface
+			continue
+		}
```

Comment:
I don't feel like this is correct... what if I have 2 providers, neither of which implements `ProviderTypeRepoLister`?

It feels like we might want to have a counter `eligibleProviders` that's incremented if the provider is a repo lister, and then return the error at the end of the loop if `eligibleProviders` is 0.